DataGrid: add the Appearance and Editing section (#5287)
DevExpress/devextreme-documentation: concepts/05 UI Components/DataGrid/20 Editing/43 Customize Edit Form/10 Appearance and Editing/20 Editor Customization.md

> DataGrid generates editors based on the **columns[]**.[dataType](/Documentation/ApiReference/UI_Components/dxDataGrid/Configuration/columns/#dataType) property. The editors are generated with the default options. This section demonstrates how to change the default editor settings, implement any of built-in editors, or create your custom editor.

Reviewer: > The editors are generated with the default options.

Does this sentence serve a particular purpose in this paragraph? Since you mention the default editor settings in the next sentence, we can just remove this awkward thing and be done with it.

> **implement** any of **built-in** editors

What do you mean by "implement" built-in editors? If they are already built-in, how can one implement them?

```suggestion
DataGrid generates editors based on the **columns[]**.[dataType](/Documentation/ApiReference/UI_Components/dxDataGrid/Configuration/columns/#dataType) property. This section demonstrates how to change the default editor settings, configure built-in editors, or create your custom editor.
```

Author: If you specify *dataType: "number"*, DataGrid uses the TextBox **by default**. You can also configure the **built-in** SelectBox editor for this field or implement a custom editor (not a built-in for DG).

Reviewer: OK. The point is that you can implement a custom editor but not a built-in (predefined, default, you name it) editor, for it's already implemented.

> If you want to change the predefined editor properties, specify the **columns[].**[editorOptions](/Documentation/ApiReference/UI_Components/dxDataGrid/Configuration/columns/#editorOptions) property, that applies to header filter editors as well. Do not specify the **onValueChanged** handler in this property, because the internal DataGrid logic overrides it. The **columns[].formItem.**[editorOptions](/Documentation/ApiReference/UI_Components/dxForm/Item_Types/SimpleItem/#editorOptions) property applies to form items only. 

Reviewer: Punctuation.

```suggestion
If you want to change the predefined editor properties, specify the **columns[].**[editorOptions](/Documentation/ApiReference/UI_Components/dxDataGrid/Configuration/columns/#editorOptions) property that also affects header filter editors. Omit the **onValueChanged** handler in this property because the internal DataGrid implementation overrides it. The **columns[].formItem.**[editorOptions](/Documentation/ApiReference/UI_Components/dxForm/Item_Types/SimpleItem/#editorOptions) property affects form items only. 
```

> [note]Do not specify the **template** property in **formItem.editorOptions** or **columns[].editorOptions**. To implement a custom template, use the [editCellTemplate](/Documentation/ApiReference/UI_Components/dxDataGrid/Configuration/columns/#editCellTemplate) property.

Reviewer: ```suggestion
[note]Leave the **template** property in **formItem.editorOptions** or **columns[].editorOptions** unspecified. To implement a custom template, use the [editCellTemplate](/Documentation/ApiReference/UI_Components/dxDataGrid/Configuration/columns/#editCellTemplate) property instead.
```

#include common-demobutton with {
    url: "https://js.devexpress.com/Demos/WidgetsGallery/Demo/DataGrid/CustomEditors/"
}

> You can override the **onValueChanged** handler of the predefined editor and change the predefined editor type. For this purpose, handle the [onEditorPreparing](/Documentation/ApiReference/UI_Components/dxDataGrid/Configuration/#onEditorPreparing) event. Use handler parameters to determine the editor location and the **dataField** to which the editor is bound. Set the **editorName** parameter to the required DevExtreme editor component and the **editorOptions** parameter to the editor configuration object.

Reviewer: > Use handler parameters to determine the editor location and the **dataField** to which the editor is bound. Set the **editorName** parameter to the required DevExtreme editor component and the **editorOptions** parameter to the editor configuration object.

These sentences look somewhat confusing. At first, you speak of handler parameters in general when it comes to editor location, and then start to name particular parameters all of the sudden. The point is, I do not understand why editor location gets special treatment in this case.

If you wan to explain everything in detail, I would recommend you to make a list of parameters along with the corresponding explanations.  Every time you iterate more than two things in plain text, consider creating a list instead.

As I mentioned earlier on, this paragraph looks somewhat confusing in its current state.

Author: I agree with your suggestions! I think, we can specify only the following information: You can override the **onValueChanged** handler of the predefined editor and change the predefined editor type. For this purpose, use the **onEditorPreparing** property.

Reviewer: Therefore, you believe that it's not worth going into greater detail, am I right? That's ultimately your choice to make -- I have no objections here.

I believe we can stop right here:

```suggestion
You can override the **onValueChanged** handler of the predefined editor and change its type. To accomplish this goal, handle the [onEditorPreparing](/Documentation/ApiReference/UI_Components/dxDataGrid/Configuration/#onEditorPreparing) event.
```

Author: Yes, there is a code snippet with details.

                        if (someCondition(e, args)) doSomething(e, args);
                        defaultValueChangeHandler(args);
                    }
                }
            }
        }
    }
    </script>

##### React

    <!-- tab: App.js -->
    import React from 'react';

    import 'devextreme/dist/css/dx.light.css';

    import DataGrid from 'devextreme-react/data-grid';

    const overrideOnValueChanged = (e) => {
        if (e.dataField === "Notes" && e.parentType === "dataRow") {
            e.editorName = "dxTextArea";
        }
        if (e.dataField === "requiredDataField" && e.parentType === "dataRow") {
            const defaultValueChangeHandler = e.editorOptions.onValueChanged;
            e.editorOptions.onValueChanged = function(args) { // override the default handler
                if (someCondition(e, args)) doSomething(e, args);
                defaultValueChangeHandler(args);
            }
        }
    }

    export default function App() {
        return (
            <DataGrid ... 
                onEditorPreparing={overrideOnValueChanged}
            >   
            </DataGrid>
        );
    }

---

> If you need to implement a custom editor for a specific data type (array, object, etc.), use the **columns[]**.[editCellTemplate](/Documentation/ApiReference/UI_Components/dxDataGrid/Configuration/columns/#editCellTemplate) property. It is important to allow users to input or change the column value. Handle this value change and use the **setValue** template parameter to change the corresponding cell value. The **editCellTemplate** property specifies one template for all editing modes.

Reviewer: > Handle this value change and use the **setValue** template parameter to change the corresponding cell value.

Do I have to handle a specific event for that purpose? If so, why don't you name the target event explicitly?

Author: Ok! My suggestion: Handle the **onValueChanged** event and use the **setValue** template parameter to change the corresponding cell value.

Reviewer: OK. Fine by me. Thank you.
